cmd/mt-index-cat/main.go

```diff
@@ -0,0 +1,144 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/raintank/dur"
+	"github.com/raintank/metrictank/cmd/mt-index-cat/out"
+	"github.com/raintank/metrictank/idx/cassandra"
+	"gopkg.in/raintank/schema.v1"
+)
+
+func perror(err error) {
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func main() {
+
+	var addr string
+	var from string
+	var maxAge string
+	var count bool
+
+	var total int
+
+	globalFlags := flag.NewFlagSet("global config flags", flag.ExitOnError)
+	globalFlags.StringVar(&addr, "addr", "http://localhost:6060", "graphite/metrictank address")
+	globalFlags.StringVar(&from, "from", "30min", "from. eg '30min', '5h', '14d', etc. or a unix timestamp")
+	globalFlags.StringVar(&maxAge, "max-age", "6h30min", "max age (last update diff with now) of metricdefs.  use 0 to disable")
+	globalFlags.BoolVar(&count, "count", false, "print number of metrics loaded to stderr")
+
+	cassFlags := cassandra.ConfigSetup()
+
+	outputs := []string{"dump", "list", "vegeta-render", "vegeta-render-patterns"}
+
+	flag.Usage = func() {
+		fmt.Println(os.Args[0])
+		fmt.Printf("Usage:\n\n")
+		fmt.Printf("  inspect-idx [global config flags] <idxtype> [idx config flags] output \n\n")
+		fmt.Printf("global config flags:\n\n")
+		globalFlags.PrintDefaults()
+		fmt.Println()
+		fmt.Printf("idxtype: only 'cass' supported for now\n\n")
+		fmt.Printf("cass config flags:\n\n")
+		cassFlags.PrintDefaults()
+		fmt.Println()
+		fmt.Printf("output: %v\n\n\n", strings.Join(outputs, "|"))
+		fmt.Println("EXAMPLES:")
+		fmt.Println("mt-index-cat -from 60min cass -hosts cassandra:9042 -keyspace raintank list")
+	}
+
+	if len(os.Args) == 2 && (os.Args[1] == "-h" || os.Args[1] == "--help") {
+		flag.Usage()
+		os.Exit(0)
+	}
+
+	if len(os.Args) < 3 {
+		flag.Usage()
+		os.Exit(-1)
+	}
+
+	var show func(d schema.MetricDefinition)
+
+	last := os.Args[len(os.Args)-1]
+	var found bool
+	for _, output := range outputs {
+		if last == output {
+			found = true
+		}
+	}
+	if !found {
+		log.Printf("invalid output %q", last)
+		flag.Usage()
+		os.Exit(-1)
+	}
+	var cassI int
+	for i, v := range os.Args {
+		if v == "cass" {
+			cassI = i
+		}
+	}
+	if cassI == 0 {
+		log.Println("only indextype 'cass' supported")
+		flag.Usage()
+		os.Exit(1)
+	}
+
+	globalFlags.Parse(os.Args[1:cassI])
+	cassFlags.Parse(os.Args[cassI+1 : len(os.Args)-1])
+	cassandra.Enabled = true
+
+	switch os.Args[len(os.Args)-1] {
+	case "dump":
+		show = out.Dump
+	case "list":
+		show = out.List
+	case "vegeta-render":
+		show = out.GetVegetaRender(addr, from)
+	case "vegeta-render-patterns":
+		show = out.GetVegetaRenderPattern(addr, from)
+	default:
+		panic("this should never happen. we already validated the output type")
```

Comment:
I also thought of this but it just gets a bit too complicated for my taste.  especially as you need to know the values of addr and from, which you only get after parsing the args, so you either use closures which take those as args (extra complication) or you construct the map after you know the values, which means you can't get the list of the keys in the help message, so you still need a separate listing.